src/qt/walletcontroller.cpp

```diff
@@ -0,0 +1,110 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <qt/walletcontroller.h>
+
+#include <interfaces/handler.h>
+#include <interfaces/node.h>
+#include <QMutexLocker>
+#include <QThread>
+
+WalletController::WalletController(interfaces::Node& node, const PlatformStyle* platform_style, OptionsModel* options_model, QObject* parent)
+    : QObject(parent)
+    , m_node(node)
+    , m_platform_style(platform_style)
+    , m_options_model(options_model)
+{
+    m_handler_load_wallet = m_node.handleLoadWallet([this](std::unique_ptr<interfaces::Wallet> wallet) {
+        getOrCreateModel(std::move(wallet));
+    });
+
+    for (auto& wallet : m_node.getWallets()) {
+        getOrCreateModel(std::move(wallet));
+    }
+}
+
+// Not using the default destructor because not all member types definitions are
+// available in the header, just forward declared.
+WalletController::~WalletController() {}
+
+std::vector<WalletModel*> WalletController::getWalletModels() const
+{
+    QMutexLocker locker(&m_mutex);
+    std::vector<WalletModel*> wallet_models;
+    wallet_models.reserve(m_wallet_models.size());
+    for (auto it : m_wallet_models) {
+        wallet_models.push_back(it.second);
+    }
+    return wallet_models;
+}
+
+std::vector<std::string> WalletController::getWalletsAvailableToLoad() const
+{
+    QMutexLocker locker(&m_mutex);
+    std::vector<std::string> wallets;
+    std::set<std::string> loaded;
+    for (const std::string& name : m_node.listWalletDir()) {
+        if (m_wallet_models.count(name) == 0) wallets.push_back(name);
+    }
+    return wallets;
+}
+
+WalletModel* WalletController::loadWallet(const std::string& path)
+{
+    std::string error, warning;
+    WalletModel* wallet_model = getOrCreateModel(m_node.loadWallet(path, error, warning));
+    if (!wallet_model) {
+        //QMessageBox::information(this, tr("Open Wallet"), QString::fromStdString(error));
+        return nullptr;
+    }
+    if (!warning.empty()) {
+        //QMessageBox::information(this, tr("Open Wallet"), QString::fromStdString(warning));
+    }
+    return wallet_model;
+}
+
+WalletModel* WalletController::getOrCreateModel(std::unique_ptr<interfaces::Wallet> wallet)
+{
+    QMutexLocker locker(&m_mutex);
+
+    // Check if model instance already exists.
+    auto name = wallet->getWalletName();
+    auto it = m_wallet_models.find(name);
+    if (it != m_wallet_models.end()) return it->second;
+
+    // Instantiate model and register it.
+    WalletModel* wallet_model = new WalletModel(std::move(wallet), m_node, m_platform_style, m_options_model, nullptr);
+    it = m_wallet_models.emplace(name, wallet_model).first;
+
+    connect(wallet_model, &WalletModel::unload, [this, it, wallet_model] {
+        // Unregister wallet model and update current if necessary.
+        {
+            QMutexLocker locker(&m_mutex);
```

Comment:
Avoid shadowing `locker` in the outer scope :-)

Reply:
Not shadowed, the above is not captured.